device-api/src/topology/mod.rs

```diff
@@ -0,0 +1,349 @@
+#![allow(warnings)]
+use std::collections::BTreeMap;
+
+use crate::topology::hwloc::{Hwloc, HwlocTopology};
+use crate::topology::hwloc_binding::*;
+use crate::topology::LinkType::*;
+use crate::{Device, DeviceResult};
+
+mod build;
+mod hwloc;
+mod hwloc_binding;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum LinkType {
+    // LinkTypeUnknown unknown
+    LinkTypeUnknown = 0,
+    // LinkTypeInterconnect two devices are connected across different cpus through interconnect.
+    LinkTypeInterconnect = 10,
+    // LinkTypeCPU two devices are connected under the same cpu, it may mean:
+    // devices are directly attached to the cpu pcie lane without PCIE switch.
+    // devices are attached to different PCIE switches under the same cpu.
+    LinkTypeCPU = 20,
+    // LinkTypeHostBridge two devices are connected under the same PCIE host bridge.
+    // Note that this does not guarantee devices are attached to the same PCIE switch.
+    // More switches could exist under the host bridge switch.
+    LinkTypeHostBridge = 30,
+
+    // NOTE(@bg): Score 40 and 50 is reserved for LinkTypeMultiSwitch and LinkTypeSingleSwitch.
+    // NOTE(@bg): Score 60 is reserved for LinkTypeBoard
+
+    // LinkTypeSoc two devices are on the same Soc chip.
+    LinkTypeSoc = 70,
+}
+
+impl LinkType {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            LinkTypeUnknown => "LinkTypeUnknown",
+            LinkTypeInterconnect => "LinkTypeInterconnect",
+            LinkTypeCPU => "LinkTypeCPU",
+            LinkTypeHostBridge => "LinkTypeHostBridge",
+            LinkTypeSoc => "LinkTypeSoc",
+        }
+    }
+}
+
+pub trait HardwareTopologyHint {
+    fn get_hw_topology_hint(&self) -> String;
+}
+
+pub struct Topology {
+    hwloc_topology: Box<dyn Hwloc>,
+    topology_matrix: BTreeMap<(String, String), LinkType>,
+}
+
+impl Topology {
+    pub fn new(devices: Vec<Device>) -> DeviceResult<Topology> {
+        let mut new_topology = Self {
+            hwloc_topology: Box::new(HwlocTopology::new()),
+            topology_matrix: BTreeMap::new(),
+        };
+
+        let keys = devices.iter().map(|d| d.busname().unwrap()).collect();
+        new_topology.populate_with_keys(keys)?;
+
+        Ok(new_topology)
+    }
+
+    fn populate_with_keys(&mut self, devices: Vec<String>) -> DeviceResult<()> {
```

Comment:
I suggest the following interface:
`fn populate_with_keys<T: Hwloc>(topology_provider: T, devices: &[Device]) -> DeviceResult<BTreeMap<(String, String), LinkType>>`

It would be nice to have aliases for repeated types.
`type TopologyInner = BTreeMap<(String, String), LinkType>`

Reply:
refactored with topology provider
eb2fd550a51a815e6c6eeebfc6b20649754a1a70